aptos-move/framework/src/natives/aggregator_natives/aggregator_snapshot.rs

```diff
@@ -0,0 +1,127 @@
+ **************************************************************************************************/
+
+fn native_snapshot(
+    context: &mut SafeNativeContext,
+    _ty_args: Vec<Type>,
+    mut args: VecDeque<Value>,
+) -> SafeNativeResult<SmallVec<[Value; 1]>> {
+    debug_assert_eq!(args.len(), 1);
+
+    context.charge(AGGREGATOR_V2_SNAPSHOT_BASE)?;
+    let (aggregator_id, _limit) = aggregator_info(&safely_pop_arg!(args, StructRef))?;
+
+    // Get aggregator.
+    let aggregator_context = context.extensions().get::<NativeAggregatorContext>();
+    let mut aggregator_data = aggregator_context.aggregator_data.borrow_mut();
+    let AggregatorSnapshotID { id } = aggregator_data.snapshot(&aggregator_id);
+
+    Ok(smallvec![Value::struct_(Struct::pack(vec![Value::u128(
+        id as u128
+    ),]))])
+}
+
+/***************************************************************************************************
+ * native fun snapshot_with_u64_limit(aggregator: &Aggregator): AggregatorSnapshot<u64>;
+ **************************************************************************************************/
+
+fn native_snapshot_with_u64_limit(
+    context: &mut SafeNativeContext,
+    _ty_args: Vec<Type>,
+    mut args: VecDeque<Value>,
+) -> SafeNativeResult<SmallVec<[Value; 1]>> {
+    debug_assert_eq!(args.len(), 1);
+
+    context.charge(AGGREGATOR_V2_SNAPSHOT_WITH_U64_LIMIT_BASE)?;
+    let (aggregator_id, _limit) = aggregator_info(&safely_pop_arg!(args, StructRef))?;
+
+    // Get aggregator.
+    let aggregator_context = context.extensions().get::<NativeAggregatorContext>();
+    let mut aggregator_data = aggregator_context.aggregator_data.borrow_mut();
+    let AggregatorSnapshotID { id } = aggregator_data.snapshot(&aggregator_id);
+
+    Ok(smallvec![Value::struct_(Struct::pack(vec![Value::u64(
+        id
+    ),]))])
+}
+
+/***************************************************************************************************
+ * native fun read_snapshot<Element>(snapshot: &AggregatorSnapshot<Element>): Element;
+ **************************************************************************************************/
+
+fn native_read_snapshot(
+    context: &mut SafeNativeContext,
+    ty_args: Vec<Type>,
+    mut args: VecDeque<Value>,
+) -> SafeNativeResult<SmallVec<[Value; 1]>> {
+    debug_assert_eq!(args.len(), 1);
+
+    context.charge(AGGREGATOR_V2_READ_SNAPSHOT_BASE)?;
+    // Get the current aggregator data.
+    let aggregator_context = context.extensions().get::<NativeAggregatorContext>();
+
+    match ty_args[0] {
+        Type::U64 => {
+            let snapshot_id = aggregator_snapshot_u64_info(&safely_pop_arg!(args, StructRef))?;
+            let aggregator_data = aggregator_context.aggregator_data.borrow_mut();
+            let value = aggregator_data.read_snapshot(AggregatorSnapshotID { id: snapshot_id });
+            assert!(
+                value <= u64::MAX as u128,
+                "Snapshot value can't exceed u64::MAX"
```

Comment:
But it can? Should we just do a check comparing to the aggregator limit?

Reply:
How can the snapshot's value exceed u64::MAX when reading AggregatorSnapshot<u64>. If the value actually does exceed u64::MAX, it's an overflow and we need to abort right?

Reply:
I meant we can have the following:
1. create an aggregator with limit > u64::MAX
2. snapshot it as snapshot it (note that the types on Move side are u64, but the value is stored in extension as u128)
3. read_snapshot_u64, which fails if value is greater than  > u64::MAX.

I think Igor suggested a check in (2) to not allow this, it is a possible solution. Otherwise we have to check here?

Reply:
Gotcha. In storage, the snapshot<u64> can only store values less than u64::MAX. If the aggregator's value exceeds u64::MAX when the snapshot was created, the read_snapshot has to abort as in the current code right?
If we instead want the abort to happen at the time of creating snapshot<u64>, then we will have to compute the aggregator's value when creating the snapshot and check if the value is below u64::MAX. I believe we want to avoid that?

Reply:
Or alternatively not allow snapshots u64 from aggregators with max value above u64::MAX

Reply:
I was suggesting (and comment in Move code refers to it), that we can cast to u64 only if max_value (i.e.limit) can be cast to u64